File: settlement/room/main/placement/UtilWallPlacability.java
package settlement.room.main.placement;

import static settlement.main.SETT.IN_BOUNDS;
import static settlement.main.SETT.JOBS;
import static settlement.main.SETT.ROOMS;
import static settlement.main.SETT.TERRAIN;

import settlement.job.Job;
import settlement.main.SETT;
import settlement.tilemap.terrain.TBuilding;
import settlement.tilemap.terrain.Terrain.TerrainTile;
import snake2d.util.map.MAP_BOOLEAN;

public final class UtilWallPlacability{
	
	private UtilWallPlacability() {
		
	}
	
	public static final MAP_BOOLEAN openingIsReal = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			return IN_BOUNDS(tx, ty) && !ROOMS().map.is(tx, ty) && get(tx, ty).roofIs();
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	public static final MAP_BOOLEAN openingCanBe = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			if (openingIsReal.is(tx, ty))
				return true;
			if (SETT.TERRAIN().MOUNTAIN.is(tx, ty))
				return placable(SETT.JOBS().clearss.tunnel,tx, ty) == null;
			else
				return placable(SETT.JOBS().build_structure.get(0).ceiling,tx, ty) == null;
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	public static final MAP_BOOLEAN openingShouldBuild = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			if (openingIsReal.is(tx, ty))
				return false;
			return openingCanBe.is(tx, ty);
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	public static final MAP_BOOLEAN wallisReal = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			return IN_BOUNDS(tx, ty) && !ROOMS().map.is(tx, ty) && get(tx, ty).isMassiveWall();
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	public static final MAP_BOOLEAN wallCanBe = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			if (wallisReal.is(tx, ty))
				return true;
			if (SETT.TERRAIN().CAVE.is(tx, ty))
				return placable(SETT.JOBS().clearss.caveFill, tx, ty) == null;
			else
				return placable(SETT.JOBS().build_structure.get(0).wall, tx, ty) == null;
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	public static final MAP_BOOLEAN wallShouldBuild = new MAP_BOOLEAN() {
		
		@Override
		public boolean is(int tx, int ty) {
			if (wallisReal.is(tx, ty))
				return false;
			return wallCanBe.is(tx, ty);
		}
		
		@Override
		public boolean is(int tile) {
			return is(tile%SETT.TWIDTH, tile/SETT.TWIDTH);
		}
	};
	
	private static TerrainTile get(int tx, int ty) {
		if (JOBS().getter.is(tx, ty))
			return JOBS().getter.get(tx, ty).becomes(tx, ty);
		return TERRAIN().get(tx, ty);
	}

	private static CharSequence placable(Job j, int tx, int ty) {
		boolean over = Job.overwrite;
		Job.overwrite = true;
		CharSequence s = j.placer().isPlacable(tx, ty, null, null);
		Job.overwrite = over;
		return s;
	}
	
	public static void wallBuild(int tx, int ty, TBuilding building) {
		
		
		if (SETT.ROOMS().map.is(tx, ty))
			return;
		if (SETT.TERRAIN().MOUNTAIN.is(tx, ty)) {
			SETT.JOBS().clearer.set(tx, ty);
			return;
		}
		if (building.wall.is(tx, ty)) {
			SETT.JOBS().clearer.set(tx, ty);
			return;
		}
		
		if (SETT.TERRAIN().CAVE.is(tx, ty)) {
			if (placable(SETT.JOBS().clearss.caveFill, tx, ty) == null)
				SETT.JOBS().clearss.caveFill.placer().place(tx, ty, null, null);
		}else if(SETT.TERRAIN().get(tx, ty) != building.wall && placable(SETT.JOBS().build_structure.get(building.structure.index()).wall, tx, ty) == null)
			SETT.JOBS().build_structure.get(building.structure.index()).wall.placer().place(tx, ty, null, null);
	}
	
	public static void openingBuild(int tx, int ty, TBuilding building) {
		if (SETT.ROOMS().map.is(tx, ty))
			return;
		if(SETT.TERRAIN().CAVE.is(tx, ty)) {
			SETT.JOBS().clearer.set(tx, ty);
			return;
		}
		if (building.roof.is(tx, ty)) {
			SETT.JOBS().clearer.set(tx, ty);
			return;
		}
		
		
		if (SETT.TERRAIN().MOUNTAIN.is(tx, ty)) {
			if (placable(SETT.JOBS().clearss.tunnel, tx, ty) == null)
				SETT.JOBS().clearss.tunnel.placer().place(tx, ty, null, null);
		}else if(placable(SETT.JOBS().build_structure.get(building.structure.index()).ceiling, tx, ty) == null)
			SETT.JOBS().build_structure.get(building.structure.index()).ceiling.placer().place(tx, ty, null, null);
	}

	
}

File: LegacySource/settlement/room/main/placement/UtilWallPlacability.cs
using System;
using System.Collections.Generic;

namespace Settlement.Room.Main.Placement
{
    public static class UtilWallPlacability
    {
        private UtilWallPlacability()
        {
        }

        public static readonly MAP_BOOLEAN openingIsReal = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                return IN_BOUNDS(tx, ty) && !ROOMS().Map.Is(tx, ty) && Get(tx, ty).RoofIs();
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        public static readonly MAP_BOOLEAN openingCanBe = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                if (openingIsReal.Is(tx, ty))
                    return true;
                if (SETT.TERRAIN().MOUNTAIN.Is(tx, ty))
                    return Placable(SETT.JOBS().Clearss.Tunnel, tx, ty) == null;
                else
                    return Placable(SETT.JOBS().BuildStructure[0].Ceiling, tx, ty) == null;
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        public static readonly MAP_BOOLEAN openingShouldBuild = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                if (openingIsReal.Is(tx, ty))
                    return false;
                return openingCanBe.Is(tx, ty);
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        public static readonly MAP_BOOLEAN wallisReal = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                return IN_BOUNDS(tx, ty) && !ROOMS().Map.Is(tx, ty) && Get(tx, ty).IsMassiveWall();
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        public static readonly MAP_BOOLEAN wallCanBe = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                if (wallisReal.Is(tx, ty))
                    return true;
                if (SETT.TERRAIN().CAVE.Is(tx, ty))
                    return Placable(SETT.JOBS().Clearss.CaveFill, tx, ty) == null;
                else
                    return Placable(SETT.JOBS().BuildStructure[0].Wall, tx, ty) == null;
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        public static readonly MAP_BOOLEAN wallShouldBuild = new MAP_BOOLEAN
        {
            Is = (tx, ty) =>
            {
                if (wallisReal.Is(tx, ty))
                    return false;
                return wallCanBe.Is(tx, ty);
            },
            IsTile = tile =>
            {
                return Is(tile % SETT.TWIDTH, tile / SETT.TWIDTH);
            }
        };

        private static TerrainTile Get(int tx, int ty)
        {
            if (JOBS().Getter.Is(tx, ty))
                return JOBS().Getter.Get(tx, ty).Becomes(tx, ty);
            return TERRAIN().Get(tx, ty);
        }

        private static string Placable(Job j, int tx, int ty)
        {
            bool over = Job.Overwrite;
            Job.Overwrite = true;
            string s = j.Placer().IsPlacable(tx, ty, null, null);
            Job.Overwrite = over;
            return s;
        }

        public static void WallBuild(int tx, int ty, TBuilding building)
        {
            if (SETT.ROOMS().Map.Is(tx, ty))
                return;
            if (SETT.TERRAIN().MOUNTAIN.Is(tx, ty))
            {
                SETT.JOBS().Clearer.Set(tx, ty);
                return;
            }
            if (building.Wall.Is(tx, ty))
            {
                SETT.JOBS().Clearer.Set(tx, ty);
                return;
            }

            if (SETT.TERRAIN().CAVE.Is(tx, ty))
            {
                if (Placable(SETT.JOBS().Clearss.CaveFill, tx, ty) == null)
                    SETT.JOBS().Clearss.CaveFill.Placer().Place(tx, ty, null, null);
            }
            else if (SETT.TERRAIN().Get(tx, ty) != building.Wall && Placable(SETT.JOBS().BuildStructure[building.Structure.Index].Wall, tx, ty) == null)
                SETT.JOBS().BuildStructure[building.Structure.Index].Wall.Placer().Place(tx, ty, null, null);
        }

        public static void OpeningBuild(int tx, int ty, TBuilding building)
        {
            if (SETT.ROOMS().Map.Is(tx, ty))
                return;
            if (SETT.TERRAIN().CAVE.Is(tx, ty))
            {
                SETT.JOBS().Clearer.Set(tx, ty);
                return;
            }
            if (building.Roof.Is(tx, ty))
            {
                SETT.JOBS().Clearer.Set(tx, ty);
                return;
            }

            if (SETT.TERRAIN().MOUNTAIN.Is(tx, ty))
            {
                if (Placable(SETT.JOBS().Clearss.Tunnel, tx, ty) == null)
                    SETT.JOBS().Clearss.Tunnel.Placer().Place(tx, ty, null, null);
            }
            else if (Placable(SETT.JOBS().BuildStructure[building.Structure.Index].Ceiling, tx, ty) == null)
                SETT.JOBS().BuildStructure[building.Structure.Index].Ceiling.Placer().Place(tx, ty, null, null);
        }
    }

    public class MAP_BOOLEAN
    {
        public Func<int, int, bool> Is { get; set; }
        public Func<int, bool> IsTile { get; set; }
    }
}